keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifact/ArtifactListener.kt

```diff
@@ -86,49 +82,65 @@ class ArtifactListener(
     if (event.controllerTriggered) {
       log.info("Fetching latest version of all registered artifacts...")
     }
-    syncDebArtifactVersions()
+    syncArtifactVersions()
   }
 
   /**
    * For each registered debian artifact, get the last version, and persist if it's newer than what we have.
    */
   // todo eb: should we fetch more than one version?
-  // todo eb: do this for docker images also
   @Scheduled(initialDelay = 60000, fixedDelayString = "\${keel.artifact-refresh.frequency:PT6H}")
-  fun syncDebArtifactVersions() =
+  fun syncArtifactVersions() =
     runBlocking {
-      artifactRepository.getAll(DEB).forEach { artifact ->
+      artifactRepository.getAll().forEach { artifact ->
         launch {
-          val lastRecordedVersion: String? = artifactRepository.versions(artifact).firstOrNull()
-          val latestVersion: String? = artifactService.getVersions(artifact.name).firstOrNull()
-          val latestAppVersion = "${artifact.name}-$latestVersion"
+          val lastRecordedVersion = getLatestStoredVersion(artifact)
+          val latestVersion = when (artifact) {
+            is DebianArtifact -> getLatestDeb(artifact)?.let { "${artifact.name}-$it" }
+            is DockerArtifact -> getLatestDockerTag(artifact)
+            else -> throw UnsupportedArtifactTypeException(artifact.type.value())
+          }
           if (latestVersion != null) {
             val hasNew = when {
               lastRecordedVersion == null -> true
-              latestAppVersion != lastRecordedVersion -> {
-                listOf(latestAppVersion, lastRecordedVersion).sortedWith(VERSION_COMPARATOR.reversed()).first() == latestAppVersion
+              latestVersion != lastRecordedVersion -> {
+                artifact.versioningStrategy.comparator.compare(lastRecordedVersion, latestVersion) > 0
               }
               else -> false
             }
 
             if (hasNew) {
               log.debug("Artifact {} has a missing version {}, persisting..", artifact, latestVersion)
-              val version = artifactService.getArtifact(artifact.name, latestVersion)
-              artifactRepository.store(artifact, latestAppVersion, debStatus(version))
+              val status = when (artifact.type) {
+                DEB -> debStatus(artifactService.getArtifact(artifact.name, latestVersion.removePrefix("${artifact.name}-")))
```

Comment:
Any reason not to use the class types in the `when` statement like you did elsewhere?

Reply:
Kotlin can't detect that there are only two valid types of artifacts from using the class type (i think because `DeliveryArtifact` isn't sealed), so I'd have to create a third `else` condition and then handle that. Seems simpler to do it this way.